### DISCUSSION TITLE
llama ops ideas

### DISCUSSION THREAD

**Discussion by iboB** [alpaca-core/alpaca-core-sdk#50] [Dev]
Currently we support a single rudimentary op: `run` which generates up to `max_tokens` tokens. 

Obviously this is not enough for a chat (or at least would make a chat clunky).

What other ops do we need?

**Comment by pminev** [comment]
Based on [that issue](https://github.com/alpaca-core/ilib-llama.cpp/issues/18) I've compiled a list of features that are used in the server.

# Llama server features

To see all supported options and server endpoints you can look at [server's documentation](https://github.com/ggerganov/llama.cpp/blob/master/examples/server/README.md). There is explained what is supported by the server and how to configure it for your use-case.

## Feature list

I'll compile a list of llama features that are used in the server and how much we've integrated them in our plugin. As a result we'll have a list of features to implement in order to be feature comparable when we're implementing our server.

### Model

| Feature | Supported by AC Llama | Prio |
| -------- | ----------- | --- |
| LLM inference | YES | - |
| Init Model with all llama_model_params | PARTIALLY | [which?](https://github.com/alpaca-core/ilib-llama.cpp/issues/20) |
| Init with KV overrides - it's in llama_model_params | NO | [low](https://github.com/alpaca-core/ilib-llama.cpp/issues/21) |
| Chat templates | YES - can be improved | [HIGH](https://github.com/alpaca-core/ilib-llama.cpp/issues/11) |
| Stream | NO - can be implemented by the client | [HIGH](https://github.com/alpaca-core/ac-local/discussions/215) |
| Antiprompts | YES | - |
| Reranking | NO | ? |
| Parallel decoding with multi-user support | YES | ? |
| Continuous batching | NO | ? |
| Multimodal | NO | ? |
| Adding LoRa | YES - it's `scaling` is considered | - | 
| Adding Control vectors | YES - it's `scaling` & `layer range` is considered | - |

### Instance specific

| Feature | Supported by AC Llama | Prio |
| -------- | ----------- | --- |
| Init with customized llama_context_params | PARTIAL| [which?](https://github.com/alpaca-core/ilib-llama.cpp/issues/22) |

### Session specific

| Feature | Supported by AC Llama | Prio |
| -------- | ----------- | --- |
| Control Sampler params (temp, top_k, top_p) | NO - there is no such session op | [HIGH](https://github.com/alpaca-core/ilib-llama.cpp/issues/23) |
| Control repeat penalty | NO | [^](https://github.com/alpaca-core/ilib-llama.cpp/issues/23) |
| Control DRY multiplier | NO | [^](https://github.com/alpaca-core/ilib-llama.cpp/issues/23) |
| Control which samplers are enabled and their order (mirostat for example) | NO | [^](https://github.com/alpaca-core/ilib-llama.cpp/issues/23) |
| Control how much last tokens to consider penalizing repeatition | NO | [^](https://github.com/alpaca-core/ilib-llama.cpp/issues/23) |
| Control response indentation (for code completion) | NO | ? |
| Control grammar-based sampling | NO | [HIGH](https://github.com/alpaca-core/ilib-llama.cpp/issues/8) |
| Should insert special tokens in the prompt (BOS, EOS ...) | NO | ? |
| Tokenize only | YES | ? |
| Should return [tokens with IDs](https://github.com/ggerganov/llama.cpp/blame/f162d45a21b27f7613f14539f9a4932d6ff3ca48/examples/server/README.md#L495) | NO | ? |
| Generate embeddings | YES | [HIGH](https://github.com/alpaca-core/ilib-llama.cpp/issues/24) |
| Generate an infill (passed prefix and suffix - for code completion)| NO | ? |
| Extract [usage analytics](https://github.com/ggerganov/llama.cpp/blame/f162d45a21b27f7613f14539f9a4932d6ff3ca48/examples/server/README.md#L799) (total tokens, kv cache usage) | NO | [low](https://github.com/alpaca-core/ilib-llama.cpp/issues/25) |



### Server specific

| Feature | Supported by AC Llama | Priority |
| -------- | ----------- | --- |
| Schema-constraned json response format | YES | - |
| [OpenAI API](https://github.com/openai/openai-openapi) compatible chat completions and embeddings routes | NO - server specific | - |
| Control prompt caching | NO | ? |
| Set a [response image data](https://github.com/ggerganov/llama.cpp/blame/f162d45a21b27f7613f14539f9a4932d6ff3ca48/examples/server/README.md#L436) | NO | ? |
| Set a [response time limit](https://github.com/ggerganov/llama.cpp/blame/f162d45a21b27f7613f14539f9a4932d6ff3ca48/examples/server/README.md#L434) | NO | ? |

** edited with prio notes **

**Comment by pminev** [comment]
`Reranking` is used to sort document relevancy according to specific prompt. In `llama.cpp`'s server you can either run inference, run only embedding or run only reranking since the models which are used are different.

OpenAI have models which can do embeddings and text-completion.

In our API this can be solved by creating new Model for each operation. Maybe we solve this at loader level by creating different Model/Instances for each type of model which supports specific operations.

**Comment by pminev** [comment]
`Parallel decoding with multi-user support` - Translated to AC language is basically support of multiple Instances for specific model which can handle multiple users.

**Comment by pminev** [comment]
`Continuous batching` - batch multiple server tasks. If there are multiple jobs for completion it'll try to batch them in single [llama decode call](https://github.com/ggerganov/llama.cpp/blob/5555c0c1f66d766c544c30699190dec0285bcbfc/examples/server/server.cpp#L2681).

**Comment by pminev** [comment]
`Multimodal` - Support of multimodal models - according to [this issue](https://github.com/ggerganov/llama.cpp/issues/8010) the support was removed and now it's WIP.

**Comment by pminev** [comment]
`Control response indentation` - in llama.cpp server this happens after `getToken` operation and verify that the indentation is minimum N whitespaces. In the terms of AC this has to be param in high-level API (kind of `completeText`) wherever it'll be implemented.

**Comment by pminev** [comment]
`Should insert special tokens in the prompt` - Should special tokens like BOS be inserted in the prompt during tokenization. It's passed to [tokenize_mixed](https://github.com/ggerganov/llama.cpp/blob/5555c0c1f66d766c544c30699190dec0285bcbfc/examples/server/utils.hpp#L96) after that to [common_tokenize](https://github.com/ggerganov/llama.cpp/blob/5555c0c1f66d766c544c30699190dec0285bcbfc/common/common.cpp#L1570)  which goes to [llama_tokenize](https://github.com/ggerganov/llama.cpp/blob/5555c0c1f66d766c544c30699190dec0285bcbfc/src/llama.cpp#L21715)

**Comment by pminev** [comment]
`Tokenize only` - The server in llama.cpp supports to make a request to give a text and to receive array of token IDs. To support it in AC we should implement such Instance op.

**Comment by pminev** [comment]
`Should return tokens with IDs` - this was incorrectly phrased. This in llama.cpp is an option of the tokenizer `with_pieces`. It means instead of returning vector of IDs to return vector of objects which contain tokenID and it's corresponding text piece.

**Comment by pminev** [comment]
`Control prompt caching` - `cache_prompt` option of `completion` request: Re-use KV cache from a previous request if possible. This way the common prefix does not have to be re-processed, only the suffix that differs between the requests. Because (depending on the backend) the logits are not guaranteed to be bit-for-bit identical for different batch sizes (prompt processing vs. token generation) enabling this option can cause nondeterministic results. Default: true
